src/Features/Core/Diagnostics/DiagnosticAnalyzerService.cs

```diff
@@ -53,10 +53,53 @@ protected DiagnosticAnalyzerService(ImmutableArray<AnalyzerReference> workspaceA
         {
             if (projectOpt == null)
             {
-                return _hostAnalyzerManager.GetHostDiagnosticDescriptorsPerReference();
+                return ConvertReferenceIdentityToName(_hostAnalyzerManager.GetHostDiagnosticDescriptorsPerReference());
             }
 
-            return _hostAnalyzerManager.CreateDiagnosticDescriptorsPerReference(projectOpt);
+            return ConvertReferenceIdentityToName(_hostAnalyzerManager.CreateDiagnosticDescriptorsPerReference(projectOpt), projectOpt);
+        }
+
+        private ImmutableDictionary<string, ImmutableArray<DiagnosticDescriptor>> ConvertReferenceIdentityToName(
+            ImmutableDictionary<object, ImmutableArray<DiagnosticDescriptor>> descriptorsPerReference, Project projectOpt = null)
+        {
+            var map = _hostAnalyzerManager.CreateAnalyzerReferencesMap(projectOpt);
+
+            var nameCollisionMap = new Dictionary<string, int>();
+            var builder = ImmutableDictionary.CreateBuilder<string, ImmutableArray<DiagnosticDescriptor>>();
+
+            foreach (var kv in descriptorsPerReference)
+            {
+                var id = kv.Key;
+                var descriptors = kv.Value;
+
+                AnalyzerReference reference;
+                if (map.TryGetValue(id, out reference) && reference != null)
+                {
+                    var displayName = GetDisplayName(nameCollisionMap, reference);
+                    builder.Add(displayName, descriptors);
+                }
+            }
+
+            return builder.ToImmutable();
+        }
+
+        private static string GetDisplayName(Dictionary<string, int> nameCollisionMap, AnalyzerReference reference)
+        {
+            var name = reference.Display ?? FeaturesResources.Unknown;
+            var count = GetNameCollisionCount(nameCollisionMap, name);
+            return count == 0 ? name : $"{name} [#{count}]";
+        }
+
```

Comment:
@heejaechang This naming is going to cause other problems with the rule set editor. Instead, if multiple `AnalyzerReferences` have the same `DisplayName` just union together their diagnostic descriptors.